### ISSUE TITLE
KubeArmor fails to initialize system monitor upon deployment

### ISSUE THREAD

**Issue by oneiro-naut** [kubearmor/KubeArmor#200]
I was trying to deploy kubearmor using yaml file in /KubeArmor/deployment/docker/kubearmor.yaml. KubeArmor terminates giving these logs. It does not happen when I run kubearmor as command. I am using the kubearmor-dev-next vagrant VM.
OS: Ubuntu 18.04.5 LTS x86_64
Kernel: 5.13.0-051300rc3-generic
```
vagrant@kubearmor-dev-next:~$ kubectl logs -n kube-system kubearmor-ktz86
2021-07-12 03:59:13.997837	INFO	Started to serve gRPC-based log feeds
2021-07-12 03:59:13.999155	INFO	Initializing an eBPF program
In file included from <built-in>:2:
In file included from /virtual/include/bcc/bpf.h:12:
In file included from include/linux/types.h:6:
In file included from include/uapi/linux/types.h:14:
In file included from ./include/uapi/linux/posix_types.h:5:
In file included from include/linux/stddef.h:5:
In file included from include/uapi/linux/stddef.h:2:
In file included from include/linux/compiler_types.h:76:
include/linux/compiler-clang.h:33:9: warning: '__HAVE_BUILTIN_BSWAP32__' macro redefined [-Wmacro-redefined]
#define __HAVE_BUILTIN_BSWAP32__
        ^
<command line>:4:9: note: previous definition is here
#define __HAVE_BUILTIN_BSWAP32__ 1
        ^
In file included from <built-in>:2:
In file included from /virtual/include/bcc/bpf.h:12:
In file included from include/linux/types.h:6:
In file included from include/uapi/linux/types.h:14:
In file included from ./include/uapi/linux/posix_types.h:5:
In file included from include/linux/stddef.h:5:
In file included from include/uapi/linux/stddef.h:2:
In file included from include/linux/compiler_types.h:76:
include/linux/compiler-clang.h:34:9: warning: '__HAVE_BUILTIN_BSWAP64__' macro redefined [-Wmacro-redefined]
#define __HAVE_BUILTIN_BSWAP64__
        ^
<command line>:5:9: note: previous definition is here
#define __HAVE_BUILTIN_BSWAP64__ 1
        ^
In file included from <built-in>:2:
In file included from /virtual/include/bcc/bpf.h:12:
In file included from include/linux/types.h:6:
In file included from include/uapi/linux/types.h:14:
In file included from ./include/uapi/linux/posix_types.h:5:
In file included from include/linux/stddef.h:5:
In file included from include/uapi/linux/stddef.h:2:
In file included from include/linux/compiler_types.h:76:
include/linux/compiler-clang.h:35:9: warning: '__HAVE_BUILTIN_BSWAP16__' macro redefined [-Wmacro-redefined]
#define __HAVE_BUILTIN_BSWAP16__
        ^
<command line>:3:9: note: previous definition is here
#define __HAVE_BUILTIN_BSWAP16__ 1
        ^
In file included from /virtual/main.c:15:
In file included from include/net/inet_sock.h:10:
In file included from include/linux/netdevice.h:19:
In file included from include/net/net_namespace.h:33:
In file included from include/net/netns/bpf.h:7:
include/linux/bpf-netns.h:21:7: error: use of undeclared identifier 'BPF_SK_LOOKUP'; did you mean 'BPF_F_LOCK'?
        case BPF_SK_LOOKUP:
             ^~~~~~~~~~~~~
             BPF_F_LOCK
/virtual/include/bcc/bpf.h:338:2: note: 'BPF_F_LOCK' declared here
        BPF_F_LOCK      = 4, /* spin_lock-ed map_lookup/map_update */
        ^
In file included from /virtual/main.c:15:
In file included from include/net/inet_sock.h:10:
In file included from include/linux/netdevice.h:23:
In file included from include/net/netprio_cgroup.h:7:
In file included from include/linux/cgroup.h:22:
In file included from include/linux/cgroup-defs.h:17:
In file included from include/linux/bpf-cgroup.h:5:
include/linux/bpf.h:806:21: error: field has incomplete type 'enum bpf_link_type'
        enum bpf_link_type type;
                           ^
include/linux/bpf.h:806:7: note: forward declaration of 'enum bpf_link_type'
        enum bpf_link_type type;
             ^
include/linux/bpf.h:820:17: warning: declaration of 'struct bpf_link_info' will not be visible outside of this function [-Wvisibility]
                              struct bpf_link_info *info);
                                     ^
include/linux/bpf.h:1236:12: warning: declaration of 'union bpf_iter_link_info' will not be visible outside of this function [-Wvisibility]
                                        union bpf_iter_link_info *linfo,
                                              ^
include/linux/bpf.h:1242:14: warning: declaration of 'struct bpf_link_info' will not be visible outside of this function [-Wvisibility]
                                         struct bpf_link_info *info);
                                                ^
include/linux/bpf.h:1285:12: warning: declaration of 'struct bpf_link_info' will not be visible outside of this function [-Wvisibility]
                                struct bpf_link_info *info);
                                       ^
In file included from /virtual/main.c:15:
In file included from include/net/inet_sock.h:13:
In file included from include/net/sock.h:27:
In file included from include/linux/filter.h:25:
In file included from include/net/sch_generic.h:21:
include/net/flow_offload.h:316:4: error: use of undeclared identifier 'KBUILD_MODNAME'
                        NL_SET_ERR_MSG_MOD(extack, "Mixing HW stats types for actions is not supported");
                        ^
include/linux/netlink.h:89:27: note: expanded from macro 'NL_SET_ERR_MSG_MOD'
        NL_SET_ERR_MSG((extack), KBUILD_MODNAME ": " msg)
                                 ^
In file included from /virtual/main.c:15:
In file included from include/net/inet_sock.h:13:
In file included from include/net/sock.h:27:
In file included from include/linux/filter.h:25:
In file included from include/net/sch_generic.h:21:
include/net/flow_offload.h:316:4: error: expected ';' at end of declaration
include/linux/netlink.h:89:42: note: expanded from macro 'NL_SET_ERR_MSG_MOD'
        NL_SET_ERR_MSG((extack), KBUILD_MODNAME ": " msg)
                                                ^
In file included from /virtual/main.c:15:
In file included from include/net/inet_sock.h:13:
In file included from include/net/sock.h:27:
In file included from include/linux/filter.h:25:
In file included from include/net/sch_generic.h:21:
include/net/flow_offload.h:351:3: error: use of undeclared identifier 'KBUILD_MODNAME'
                NL_SET_ERR_MSG_MOD(extack, "Driver supports only default HW stats type \"any\"");
                ^
include/linux/netlink.h:89:27: note: expanded from macro 'NL_SET_ERR_MSG_MOD'
        NL_SET_ERR_MSG((extack), KBUILD_MODNAME ": " msg)
                                 ^
In file included from /virtual/main.c:15:
In file included from include/net/inet_sock.h:13:
In file included from include/net/sock.h:27:
In file included from include/linux/filter.h:25:
In file included from include/net/sch_generic.h:21:
include/net/flow_offload.h:351:3: error: expected ';' at end of declaration
include/linux/netlink.h:89:42: note: expanded from macro 'NL_SET_ERR_MSG_MOD'
        NL_SET_ERR_MSG((extack), KBUILD_MODNAME ": " msg)
                                                ^
In file included from /virtual/main.c:15:
In file included from include/net/inet_sock.h:13:
In file included from include/net/sock.h:27:
In file included from include/linux/filter.h:25:
In file included from include/net/sch_generic.h:21:
include/net/flow_offload.h:355:3: error: use of undeclared identifier 'KBUILD_MODNAME'
                NL_SET_ERR_MSG_MOD(extack, "Driver does not support selected HW stats type");
                ^
include/linux/netlink.h:89:27: note: expanded from macro 'NL_SET_ERR_MSG_MOD'
        NL_SET_ERR_MSG((extack), KBUILD_MODNAME ": " msg)
                                 ^
In file included from /virtual/main.c:15:
In file included from include/net/inet_sock.h:13:
In file included from include/net/sock.h:27:
In file included from include/linux/filter.h:25:
In file included from include/net/sch_generic.h:21:
include/net/flow_offload.h:355:3: error: expected ';' at end of declaration
include/linux/netlink.h:89:42: note: expanded from macro 'NL_SET_ERR_MSG_MOD'
        NL_SET_ERR_MSG((extack), KBUILD_MODNAME ": " msg)
                                                ^
7 warnings and 8 errors generated.
2021-07-12 03:59:17.320643	ERROR	Failed to initialize the system monitor
github.com/kubearmor/KubeArmor/KubeArmor/log.Err
	/usr/src/KubeArmor/KubeArmor/log/logger.go:99
github.com/kubearmor/KubeArmor/KubeArmor/feeder.(*Feeder).Err
	/usr/src/KubeArmor/KubeArmor/feeder/feeder.go:423
github.com/kubearmor/KubeArmor/KubeArmor/core.KubeArmor
	/usr/src/KubeArmor/KubeArmor/core/kubeArmor.go:354
main.main
	/usr/src/KubeArmor/KubeArmor/main.go:60
runtime.main
	/usr/local/go/src/runtime/proc.go:204
2021-07-12 03:59:17.320749	INFO	Stopped the system monitor
2021-07-12 03:59:17.320753	INFO	Terminated the KubeArmor
2021-07-12 03:59:19.331245	INFO	Stopped the log feeder
2021-07-12 03:59:19.331580	INFO	Waiting for remaining routine terminations
```

**Comment by nam-jaehyun** [comment]
@oneiro-naut does it mean that KubeArmor has some eBPF issues in the latest Kernel?

**Comment by oneiro-naut** [comment]
@nam-jaehyun It might be the case. I am going to check if something similar is happening on kernel: 4.15 vagrant VM to make sure. Strange thing here is that when we run kubearmor using `make run` on master branch it doesn't fail. It should fail giving similar error in that case as well?

**Comment by oneiro-naut** [comment]
> @oneiro-naut does it mean that KubeArmor has some eBPF issues in the latest Kernel?

@nam-jaehyun Yes. It's happening only in the latest kernel. I checked on usual kubearmor-dev VM(kernel 4.15) in that it can be deployed just fine.

**Comment by nam-jaehyun** [comment]
Related issue: https://github.com/iovisor/bcc/issues/3255

**Comment by nam-jaehyun** [comment]
This issue occurred due to the older version (0.20.0) of bcc, which is supported by Alpine.
Recently, bcc started to support Linux kernel 5.13 in the latest one (0.21.0).
Once we have the newer bcc version (0.21.0) from Alpine packages, it would be solved.

**Comment by oneiro-naut** [comment]
> Recently, bcc started to support Linux kernel 5.13 in the latest one (0.21.0).

@nam-jaehyun The same issue is happening on kernel version 5.8(on ubuntu 20.04) as tested by @weirdwiz.

**Comment by elfadel** [comment]
> > Recently, bcc started to support Linux kernel 5.13 in the latest one (0.21.0).
> 
> @nam-jaehyun The same issue is happening on kernel version 5.8(on ubuntu 20.04) as tested by @weirdwiz.

+1. Not with KubeArmor. But I am facing the same issue when compiling the [entrypoint-handler](https://github.com/elfadel/entrypoint-handler) with Ubuntu 20.10 on kernel v5.8.

**Comment by weirdwiz** [comment]
> > Recently, bcc started to support Linux kernel 5.13 in the latest one (0.21.0).
> 
> @nam-jaehyun The same issue is happening on kernel version 5.8(on ubuntu 20.04) as tested by @weirdwiz.

We do not have a problem with `BPF_SK_LOOKUP` on 5.8. And if we define `BPF_SK_LOOKUP` on 5.13. We only get an error regarding `bpf_link_type`

Relevant log for 5.8:
``` 
In file included from <built-in>:2:
In file included from /virtual/include/bcc/bpf.h:12:
In file included from include/linux/types.h:6:
In file included from include/uapi/linux/types.h:14:
In file included from ./include/uapi/linux/posix_types.h:5:
In file included from include/linux/stddef.h:5:
In file included from include/uapi/linux/stddef.h:2:
In file included from include/linux/compiler_types.h:63:
include/linux/compiler-clang.h:37:9: warning: '__HAVE_BUILTIN_BSWAP32__' macro redefined [-Wmacro-redefined]
#define __HAVE_BUILTIN_BSWAP32__
        ^
<command line>:4:9: note: previous definition is here
#define __HAVE_BUILTIN_BSWAP32__ 1
        ^
In file included from <built-in>:2:
In file included from /virtual/include/bcc/bpf.h:12:
In file included from include/linux/types.h:6:
In file included from include/uapi/linux/types.h:14:
In file included from ./include/uapi/linux/posix_types.h:5:
In file included from include/linux/stddef.h:5:
In file included from include/uapi/linux/stddef.h:2:
In file included from include/linux/compiler_types.h:63:
include/linux/compiler-clang.h:38:9: warning: '__HAVE_BUILTIN_BSWAP64__' macro redefined [-Wmacro-redefined]
#define __HAVE_BUILTIN_BSWAP64__
        ^
<command line>:5:9: note: previous definition is here
#define __HAVE_BUILTIN_BSWAP64__ 1
        ^
In file included from <built-in>:2:
In file included from /virtual/include/bcc/bpf.h:12:
In file included from include/linux/types.h:6:
In file included from include/uapi/linux/types.h:14:
In file included from ./include/uapi/linux/posix_types.h:5:
In file included from include/linux/stddef.h:5:
In file included from include/uapi/linux/stddef.h:2:
In file included from include/linux/compiler_types.h:63:
include/linux/compiler-clang.h:39:9: warning: '__HAVE_BUILTIN_BSWAP16__' macro redefined [-Wmacro-redefined]
#define __HAVE_BUILTIN_BSWAP16__
        ^
<command line>:3:9: note: previous definition is here
#define __HAVE_BUILTIN_BSWAP16__ 1
        ^
In file included from /virtual/main.c:28:
In file included from include/net/inet_sock.h:19:
In file included from include/linux/netdevice.h:42:
In file included from include/net/netprio_cgroup.h:11:
In file included from include/linux/cgroup.h:28:
In file included from include/linux/cgroup-defs.h:22:
In file included from include/linux/bpf-cgroup.h:5:
include/linux/bpf.h:1127:21: error: field has incomplete type 'enum bpf_link_type'
        enum bpf_link_type type;
                           ^
include/linux/bpf.h:1127:7: note: forward declaration of 'enum bpf_link_type'
        enum bpf_link_type type;
             ^
include/linux/bpf.h:1147:17: warning: declaration of 'struct bpf_link_info' will not be visible outside of this function [-Wvisibility]
                              struct bpf_link_info *info);
```

**Comment by oneiro-naut** [comment]
@nam-jaehyun Upgrading alpine version(we still use 3.12 and that has bcc version 0.14.0) will fix this. Currently our kubearmor image has very old bcc version in it. The same bcc and alpine is being used for all kernels. 
```
vagrant@kubearmor-dev:~$ kubectl exec -ti -n kube-system kubearmor-5g9ws -- bash
bash-5.0# apk list bcc
bcc-0.14.0-r0 x86_64 {bcc} (Apache-2.0) [installed]
```
Changin alpine version to 3.14 will fix this(since 3.14 has bcc 0.20.0 and these issues are fixed for that release):
bcc version 0.20.0 has these related points in its release notes:
1. Fix possible marco HAVE_BUILTIN_BSWAP redefine warning for kernel >= 5.10.(which is the same error we see here)
2. Support for kernel up to 5.12
@weirdwiz is currently testing these changes on ubuntu 20.04.

EDIT:
~~Image update wasn't needed. We just needed to switch to edge branch for bcc.~~
